### PR TITLE
feat(cast): cast send tries to decode custom errors

### PR DESCRIPTION
**Pull request by rplusq**
Co-authored-by: Howard <hwrdtm@users.noreply.github.com>

## Motivation

Even though issue #8606 is partially solved, this would be an improvement over the UX. Only applies this for custom errors

Nightly cast send:

```
cast send 0x5FbDB2315678afecb367f032d93F642f64180aa3 "setValue(uint256)" 101 --private-key=0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80
Error:
server returned an error response: error code 3: execution reverted: custom error 7a0e1985:ed, data: "0x7a0e198500000000000000000000000000000000000000000000000000000000000000650000000000000000000000000000000000000000000000000000000000000064"
```

Our PR:
```
cast send 0x5FbDB2315678afecb367f032d93F642f64180aa3 "setValue(uint256)" 101 --private-key=0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80

Error: 
Reverted with custom error: 
 Possible methods:
 - ValueTooHigh(uint256,uint256)
 ------------
 [000]: 0000000000000000000000000000000000000000000000000000000000000065
 [020]: 0000000000000000000000000000000000000000000000000000000000000064


Context:
- server returned an error response: error code 3: execution reverted: custom error 7a0e1985:ed, data: "0x7a0e198500000000000000000000000000000000000000000000000000000000000000650000000000000000000000000000000000000000000000000000000000000064"

```
## Solution

Using pretty_calldata to decode the custom error.

### CODE REVIEW

**Comment by rplusq** [comment]
Thinking it solves https://github.com/foundry-rs/foundry/issues/8603 as well

**Review by ArkFunds — CHANGES_REQUESTED** [review]

**Review by zerosnacks — DISMISSED** [review]
Thanks for your PR! LGTM

Confirming previously:

```
cast send 0x5FbDB2315678afecb367f032d93F642f64180aa3 "setValue(uint256)" 101 --private-key=0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80
Error: 
server returned an error response: error code 3: execution reverted: custom error 0x7a0e1985: ed, data: "0x7a0e198500000000000000000000000000000000000000000000000000000000000000650000000000000000000000000000000000000000000000000000000000000064"
```

```
cast send 0x5FbDB2315678afecb367f032d93F642f64180aa3 "setValue(uint256)" 101 --private-key=0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80
Error: 
Reverted with custom error: 
 Possible methods:
 - ValueTooHigh(uint256,uint256)
 ------------
 [000]: 0000000000000000000000000000000000000000000000000000000000000065
 [020]: 0000000000000000000000000000000000000000000000000000000000000064


Context:
- server returned an error response: error code 3: execution reverted: custom error 0x7a0e1985: ed, data: "0x7a0e198500000000000000000000000000000000000000000000000000000000000000650000000000000000000000000000000000000000000000000000000000000064"
```

**Review by grandizzy — COMMENTED** [review]
I think would worth adding a test similar with this one which starts anvil
https://github.com/foundry-rs/foundry/blob/ccabf8b71831ed872988bba0137fb760b18dc519/crates/cast/tests/cli/main.rs#L1290-L1294

, deploys contract, calls a fn that always revert with a custom error and make sure output is expected
https://github.com/foundry-rs/foundry/blob/ccabf8b71831ed872988bba0137fb760b18dc519/crates/cast/tests/cli/main.rs#L1308-L1312

Since https://github.com/foundry-rs/foundry/issues/8603 is mentioned, I see in issue description that reporter says custom error used to be displayed in old version of cast, so wonder if it's not only a regression and could avoid fetching signatures from openchain...

**Comment by rplusq** [comment]
> Since #8603 is mentioned, I see in issue description that reporter says custom error used to be displayed in old version of cast, so wonder if it's not only a regression and could avoid fetching signatures from openchain...

Yep, that was eventually fixed. The addition to fetch signatures from openchain I added to make those more readable when possible. It's opinionated of course, but I think it improves the ux.

**Comment by rplusq** [comment]
@grandizzy any idea why my test keeps failing? Seems like something is missing and I cannot figure it out

CleanShot 2024-09-25 at 22 38 21@2x

**Comment by grandizzy** [comment]
> @grandizzy any idea why my test keeps failing? Seems like something is missing and I cannot figure it out
> 
you can run test with SNAPSHOTS=override but since that's not relevant for test you could just redact as
```rust
        .stderr_eq(str![[r#"
Error: 
Reverted with custom error: 
 Possible methods:
 - ValueTooHigh(uint256,uint256)
 ------------
 [000]: 0000000000000000000000000000000000000000000000000000000000000065
 [020]: 0000000000000000000000000000000000000000000000000000000000000064
...
"#]]);
```

**Comment by rplusq** [comment]
Fixed, thanks @grandizzy

**Comment by rplusq** [comment]
Not sure if this PR will solve #8348. My PR is currently limited to cast send, though would be interested in making it more generic for the rest of the stack.

**Comment by jenpaff** [comment]
> Not sure if this PR will solve #8348. My PR is currently limited to cast send, though would be interested in making it more generic for the rest of the stack.
> 
> 
> 
> cc: @jenpaff 

Thanks for flagging, let me reopen then & re-evaluate

File: crates/cast/tests/cli/main.rs [hunk 1]
```diff
@@ -1411,6 +1411,71 @@ casttest!(block_number_hash, |_prj, cmd| {
...
+    // Deploy the contract
+    let output = cmd
+        .forge_fuse()
+        .args([
+            "create",
+            "./src/SimpleStorage.sol:SimpleStorage",
+            "--rpc-url",
+            &endpoint,
+            "--private-key",
+            "0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80",
```
Comment (zerosnacks) [inline crates/cast/tests/cli/main.rs:1445]:
```suggestion
        .args([
            "create",
            "./src/SimpleStorage.sol:SimpleStorage",
            "--rpc-url",
            &endpoint,
            "--broadcast",
            "--private-key",
            "0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80",
```

we've recently added a requirement for passing the `--broadcast` flag to `forge create`, hence the failing test

Comment (rplusq) [reply crates/cast/tests/cli/main.rs:1445]: Solved, thanks

**Comment by zerosnacks** [comment]
Hi @rplusq,

My apologies for the delay. The reason why the PR has been open for some time is because there has been an ongoing discussion around what the PR is attempting to achieve and whether there are better ways to address the core issue.

We've decided not to move ahead with the approach of this PR. Instead we've added `cast decode-error` in https://github.com/foundry-rs/foundry/pull/9428 which handles the resolving of custom errors by either `--sig`, local cache and the `openchain API`

The proposed way of enabling what this PR is attempting to achieve is by piping the output of `cast send` through `cast decode-error` as follows (or manually):

```
$ forge selectors cache

$ cast send 0x5FbDB2315678afecb367f032d93F642f64180aa3 "setValue(uint256)" 101 --private-key=0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80 2>&1 | grep -oP '(?<=data: ")0x[0-9a-fA-F]+' | xargs -I{} cast decode-error {}'
```

It was not an easy decision considering the time and effort you've spend on this PR and for that we apologize. For the future we aim for quicker feedback loops.

**Pull request closed without merging**